tests/integration/test_boltdb_v11_to_tsdb_v12_migration.py

```diff
@@ -0,0 +1,113 @@
+    await deploy_charm_from_charmhub_v11(ops_test, V11_APP_NAME)
+    await upgrade_charm_with_local_charm_v12(ops_test, V11_APP_NAME, loki_charm)
+    await verify_upgrade_success(ops_test, V11_APP_NAME, False)
+
+    # Here we upgrade again to ensure config is persisted and won't be overwritten with any weird values
+    await upgrade_charm_with_local_charm_v12(ops_test, V11_APP_NAME, loki_charm)
+    await verify_upgrade_success(ops_test, V11_APP_NAME, False)
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_and_upgrade_v12_locally(ops_test: OpsTest, loki_charm):
+    """Deploy from a local charm (v12 schema) and upgrade locally."""
+    await deploy_local_charm_v12(ops_test, V12_APP_NAME, loki_charm)
+    await upgrade_charm_with_local_charm_v12(ops_test, V12_APP_NAME, loki_charm)
+    await verify_upgrade_success(ops_test, V12_APP_NAME, True)
+
+    # Here we upgrade again to ensure config is persisted and won't be overwritten with any weird values
+    await upgrade_charm_with_local_charm_v12(ops_test, V12_APP_NAME, loki_charm)
+    await verify_upgrade_success(ops_test, V12_APP_NAME, True)
+
+
+async def deploy_charm_from_charmhub_v11(ops_test: OpsTest, app_name):
+    """Deploy the charm from Charmhub."""
+    logger.debug("Deploying charm from Charmhub")
+    await ops_test.model.deploy(
+        "ch:loki-k8s",
+        application_name=app_name,
+        channel="edge",
+        revision=140,
+        trust=True,
+    )
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
+
+
+async def deploy_local_charm_v12(ops_test: OpsTest, app_name, loki_charm):
+    """Deploy the charm-under-test."""
+    logger.debug("deploy local charm")
+    await ops_test.model.deploy(
+        loki_charm, application_name=app_name, resources=resources, trust=True
+    )
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
+
+
+async def upgrade_charm_with_local_charm_v12(ops_test: OpsTest, app_name, loki_charm):
+    """Upgrade the deployed charm with the local charm."""
+    logger.debug("Upgrading deployed charm with local charm %s", loki_charm)
+    await ops_test.model.applications[app_name].refresh(path=loki_charm, resources=resources)
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
+
+
+async def verify_upgrade_success(ops_test: OpsTest, app_name, is_fresh_installation: bool):
+    """Verify that the upgrade was successful."""
+    assert await is_loki_up(ops_test, app_name)
+    post_upgrade_config = await loki_config(ops_test, app_name)
+    tsdb_shipper_config = post_upgrade_config["storage_config"]["tsdb_shipper"]
+    tsdb_schema_configs = post_upgrade_config["schema_config"]["configs"][1]
+
+    expected_from = (
+        (datetime.date.today()).strftime("%Y-%m-%d")
+        if is_fresh_installation
+        else (datetime.date.today() + datetime.timedelta(days=1)).strftime("%Y-%m-%d")
+    )
+
+    assert all(
```

Comment:
Iirc, in a test failure it would be easier to figure out what's wrong if we use individual assert statements.